indigo/indigo/src/main/scala/indigo/shared/platform/DisplayObjectConversions.scala

```diff
@@ -146,24 +152,50 @@ final class DisplayObjectConversions(
       cloneData = batch.uniformBlocks.toJSArray.map(uniformDataConvert)
     )
 
-  def sceneNodesToDisplayObjects(
-      sceneNodes: List[SceneGraphNode],
+  def processSceneNodes(
+      sceneNodes: scalajs.js.Array[SceneNode],
       gameTime: GameTime,
       assetMapping: AssetMapping,
-      cloneBlankDisplayObjects: => HashMap[CloneId, DisplayObject],
+      cloneBlankDisplayObjects: => scalajs.js.Dictionary[DisplayObject],
       renderingTechnology: RenderingTechnology,
-      maxBatchSize: Int
-  ): (scalajs.js.Array[DisplayEntity], scalajs.js.Array[(CloneId, DisplayObject)]) =
+      maxBatchSize: Int,
+      inputEvents: => scalajs.js.Array[GlobalEvent],
+      sendEvent: GlobalEvent => Unit
+  ): (scalajs.js.Array[DisplayEntity], scalajs.js.Array[(String, DisplayObject)]) =
     val f =
       sceneNodeToDisplayObject(
         gameTime,
         assetMapping,
         cloneBlankDisplayObjects,
         renderingTechnology,
-        maxBatchSize
+        maxBatchSize,
+        inputEvents,
+        sendEvent
       )
-    val l = sceneNodes.toJSArray.map(f)
-    (l.map(_._1), l.foldLeft(scalajs.js.Array[(CloneId, DisplayObject)]())(_ ++ _._2))
+
+    val l = sceneNodes.map { node =>
```

Comment:
To further optimize this, I think you could `foldLeft` here instead of mapping and then folding later.

Reply:
I think you're right - equally I'm disinclined to get into it right now - please feel free to have a go! (...and if you don't I'll try and do so next time I'm in the area.)